### PR TITLE
[#120] Add repoCabalUrl field to Repo type

### LINKED ISSUE #120: Add `repoDefaultBranch` field to `Repo` type
We are downloading `cabal` files for repos. In order to get the correct `cabal` file, we need to know the default branch of the repo. We can get the default branch of the repo via the `github` library. We could call the `github` library each time we need the default branch (each time we download a `cabal` file), but we would go over the rate limit of the API. Instead, we could add a `repoDefaultBranch` to the `Repo` type and use it when needed. We would construct the repos the way we do now, but we just add `repoDefaultBranch`. If we are going to store the default branch we might as well store the URL for the `cabal` file for each repo. This is my logic for adding `repoCabalUrl` to the `Repo` type. Our function `fromGitHubRepo :: GitHub.Repo -> Repo` would get the `defaultBranch` and we would construct the `Repo` with this field. The thing is `defaultBranch` on `GitHub.Repo` returns a `Maybe Text`:
https://hackage.haskell.org/package/github-0.22/docs/GitHub-Data-Repos.html
If it returns nothing, the default branch should be `master`. We would have to modify a few functions, but not much.

Refer to this comment for more about the reason for this issue:
https://github.com/kowainik/issue-wanted/pull/119#discussion_r302806720

### PR DESCRIPTION
**Pull request by rashadg1030**
Resolves #120

### CODE REVIEW

**Comment by chshersh** [comment]
I think that this PR definitely can be implemented separately from the `sync` worker. Should be an independent change :slightly_smiling_face:

**Comment by rashadg1030** [comment]
I branched all of these PRs from branch #127 because I thought it was going to be merged first, and then these could rebase on top of it nicely, but in hindsight that wasn't a good idea...

**Comment by rashadg1030** [comment]
@chshersh @vrom911 Should I close this and then reopen?

**Comment by chshersh** [comment]
@rashadg1030 You can choose whatever approach for this PR which is more convenient. You can close this PR and open a new one with only relevant changes. Or you can try to prettify the history in this PR by dropping irrelevant commits.

**Comment by rashadg1030** [comment]
@chshersh I dropped commits that weren't relevant to this branch. This should be good to go.

File: src/IW/Core/Repo.hs [hunk 1]
```diff
@@ -33,12 +36,35 @@ data Repo = Repo
...
+
+    name :: RepoName
+    name  = RepoName $ GitHub.untagName $ GitHub.repoName githubRepo
+
+-- | This function returns a @Url@ for downloading a @Repo@'s @.cabal@ file.
+mkRepoCabalUrl :: RepoOwner -> RepoName -> Maybe Text -> Url
+mkRepoCabalUrl (RepoOwner repoOwner) (RepoName repoName) maybeBranch = Url $
+    "https://raw.githubusercontent.com/"
+    <> repoOwner
+    <> "/"
```
Comment (vrom911) [inline src/IW/Core/Repo.hs]: You can use `T.intercalate "/"` to not repeat slashes, but this is just the matter of taste :)

File: src/IW/Sync/Search.hs [hunk 2]
```diff
@@ -11,24 +11,31 @@ module IW.Sync.Search
...
        , searchAllHaskellIssues
        , searchAllHaskellIssuesByLabels
-       , parseUserData
        ) where
 
 import Data.Time (Day (..), addDays)
-import GitHub (SearchResult (..), URL (..), RateLimit (..), Limits, Paths, QueryString, executeRequest', limitsRemaining)
+import GitHub ( GenRequest (..)
+              , SearchResult (..)
+              , RateLimit (..)
```
Comment (chshersh) [inline src/IW/Sync/Search.hs]: This import is formatted in a weird way... Are you using `stylish-haskell` with the config from this repo?

Comment (rashadg1030) [reply src/IW/Sync/Search.hs]: @chshersh No, I'm not using `stylish-haskell` at all. This was me. I will fix this.

File: src/IW/Core/Repo.hs [hunk 3]
```diff
@@ -33,12 +36,32 @@ data Repo = Repo
...
+    owner = RepoOwner $ GitHub.untagName $ GitHub.simpleOwnerLogin $ GitHub.repoOwner githubRepo
+
+    name :: RepoName
+    name  = RepoName $ GitHub.untagName $ GitHub.repoName githubRepo
+
+-- | This function returns a @Url@ for downloading a @Repo@'s @.cabal@ file.
+mkRepoCabalUrl :: RepoOwner -> RepoName -> Maybe Text -> Url
+mkRepoCabalUrl (RepoOwner repoOwner) (RepoName repoName) maybeBranch = Url $
+    "https://raw.githubusercontent.com/"
+    <> Text.intercalate "/" [ repoOwner
```
Comment (chshersh) [inline src/IW/Core/Repo.hs]: Personal preference, but to may taste this looked better like this:

```haskell
Text.intercalate "/"
    [ "https://raw.githubusercontent.com"
    , repoOwner
    , ...
    ]
```

File: src/IW/Core/Repo.hs [hunk 4]
```diff
@@ -33,12 +36,33 @@ data Repo = Repo
...
+    owner = RepoOwner $ GitHub.untagName $ GitHub.simpleOwnerLogin $ GitHub.repoOwner githubRepo
+
+    name :: RepoName
+    name  = RepoName $ GitHub.untagName $ GitHub.repoName githubRepo
+
+-- | This function returns a @Url@ for downloading a @Repo@'s @.cabal@ file.
+mkRepoCabalUrl :: RepoOwner -> RepoName -> Maybe Text -> Url
+mkRepoCabalUrl (RepoOwner repoOwner) (RepoName repoName) maybeBranch = Url $
+    "https://raw.githubusercontent.com/"
+    <> Text.intercalate "/"
```
Comment (chshersh) [inline src/IW/Core/Repo.hs]: I was mostly telling about putting the hardcoded `https://...` string inside the list and merging `interclate` line with `Url $ ` line to have more consistent code, not use `<>` and have less lines

Comment (rashadg1030) [reply src/IW/Core/Repo.hs]: @chshersh I missed that :man_facepalming: That's definitely better

**Comment by chshersh** [comment]
The build is failing in this branch...

**Comment by rashadg1030** [comment]
I think I messed up my local branch history or something, but when I rebase onto `master` it tells me that everything is up to date and there are no conflicts when they obviously are. Not sure how to fix. I've looked online for some help, but not working. I'm going to work on other branches and then once one of those is merged, I'll try to rebase again and then maybe the conflicts will show up.

**Comment by rashadg1030** [comment]
@chshersh @vrom911 The thing is, the build is failing because the tests aren't the right type on this branch for 'parseUserData`. The master branch has the correct tests for `parsUserData` so if I change the tests for `parseUserData` on this branch and then it merges with the master branch, it will overwrite the correct tests on `master`. When I rebase I expected to see conflicts on a few files, but it allows me to rebase successfully which shouldn't be correct. I think it's because I started this new branch of a feature branch instead of master and now git thinks there are no conflicts? Not sure...I've learned my lesson and I'm just going to stick to branching from `master`. I could just cherry-pick the code myself or maybe start a new branch. I'm probably going to do the latter.

**Comment by chshersh** [comment]
@rashadg1030 Will it work if you just copy-paste the tests from `master`?

**Comment by rashadg1030** [comment]
Yes...I didn’t think of that. I went to go get something to eat but I will
try that. That should fix it. Thank you x100

**Comment by rashadg1030** [comment]
@chsersh okay actually that won’t work. I think I’m gonna just start a new branch. After looking at the code, there a lot more differences than just the tests...The repo type and a few other things are different. It’s weird that I can rebase on master without conflicts. Probably just gonna open a fresh branch and make changes from there.

**Comment by rashadg1030** [comment]
@chshersh @vrom911 This PR is finally good to go I think! I was able to rebase everything correctly and it looks good. Sorry if the commit history is messed up. I know the git commit history is important and this is really something I need to work on.

**Comment by chshersh** [comment]
@rashadg1030 Okay, frontend is merged now. So let's rebase this small PR one last time and merge it to `master` :+1: 

Regarding commit history: don't worry about that! Actually, I'm okay with having any mess inside a PR or a branch. We do `Squash and merge` anyways, so it doesn't matter how local history for the branch looked like. I'm also reviewing PRs thoroughly after every change, so the shape of history doesn't affect my reviewing workflow.

**Pull request merged**